### ISSUE TITLE
Command line arguments that contain '='

### ISSUE THREAD

**Issue by rfleschenberg** [ansible/ansible#9899]
There does not seem to be a way to pass command line arguments that contain `=` to modules:

```
- synchronize: 
    src: "{{ src }}"
    dest: "{{ dst }}"
    rsync_opts: --chmod 'u=rwX,g=rX'
```

Despite the quoting, Ansible splits on the `=`, resulting in this on the command line:

```
--chmod 'u=rwX g=rX'
```

The same is true if you use Ansible's key=value syntax:

```
- synchronize: src="{{ pelican_src_dir }}"
               dest="{{ pelican_dst_dir }}"
               rsync_opts=--chmod='u=rwX,g=rX'
```

**Comment by rfleschenberg** [comment]
I also verified this against the current git master branch.

**Comment by jimi-c** [comment]
Hi, this situation is no longer present in the devel branch (which will be ansible 2.0). Here is the output using a version of your example above:

<pre>
# cat 9899.yml 
- hosts: localhost
  gather_facts: no
  tasks:
  - synchronize: 
      src: "9899_testing/"
      dest: "/tmp/9899_1/"
      rsync_opts: --chmod='ugo=rwX'

  - synchronize: src="9899_testing/" dest="/tmp/9899_2/" rsync_opts="--chmod='ugo=rwX'"


PLAY: ***************************************************************************

TASK [synchronize dest=/tmp/9899_1/ src=9899_testing/ rsync_opts=--chmod='ugo=rwX'] ***
changed: [localhost] => {"changed": true, "cmd": "rsync --delay-updates -F --compress --archive --rsh 'ssh  -S none -o StrictHostKeyChecking=no' --chmod='ugo=rwX' --out-format='<<CHANGED>>%i %n%L' \"9899_testing/\" \"/tmp/9899_1/\"", "msg": ".d..t...... ./\n>f+++++++++ a\n>f+++++++++ b\n>f+++++++++ c\n", "rc": 0, "invocation": {"module_name": "synchronize", "module_args": {"dest": "/tmp/9899_1/", "src": "9899_testing/", "rsync_opts": "--chmod='ugo=rwX'"}}, "stdout_lines": [".d..t...... ./", ">f+++++++++ a", ">f+++++++++ b", ">f+++++++++ c"]}

TASK [synchronize dest=/tmp/9899_2/ src=9899_testing/ rsync_opts=--chmod='ugo=rwX'] ***
changed: [localhost] => {"changed": true, "cmd": "rsync --delay-updates -F --compress --archive --rsh 'ssh  -S none -o StrictHostKeyChecking=no' --chmod='ugo=rwX' --out-format='<<CHANGED>>%i %n%L' \"9899_testing/\" \"/tmp/9899_2/\"", "msg": ".d..t...... ./\n>f+++++++++ a\n>f+++++++++ b\n>f+++++++++ c\n", "rc": 0, "invocation": {"module_name": "synchronize", "module_args": {"dest": "/tmp/9899_2/", "src": "9899_testing/", "rsync_opts": "--chmod='ugo=rwX'"}}, "stdout_lines": [".d..t...... ./", ">f+++++++++ a", ">f+++++++++ b", ">f+++++++++ c"]}

PLAY RECAP **********************************************************************
localhost                  : ok=2    changed=2    unreachable=0    failed=0   

</pre>

Using your permissions, rsync threw an error (`"msg": "rsync: Invalid argument passed to --chmod (u=rwX g=rX)\nrsync error: syntax or usage error (code 1) at main.c(1554) [client=3.1.1]\n"`), but with the above playbook there was no issue.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:

   * https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
   * https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by rfleschenberg** [comment]
`rsync --chmod u=rwX,g=rX /tmp/foo /tmp/bar` runs fine in a shell here. If you get an error with my playbook, either you have a different rsync version, or this issue is not solved completely yet. I notice that your playbook does not contain an argument that contains a comma. Is there maybe still a problem with passing arguments that contain commas?

**Comment by jimi-c** [comment]
With the exact above example, this is still failing as follows:

<pre>
TASK [synchronize dest=/tmp/9899/ src=roles/ rsync_opts=--chmod='u=rwX,g=rX'] ***
fatal: [localhost]: FAILED! => {"changed": false, "cmd": "rsync --delay-updates -F --compress --archive --rsh 'ssh  -S none -o StrictHostKeyChecking=no' --chmod='u=rwX g=rX' --out-format='<<CHANGED>>%i %n%L' \"roles/\" \"/tmp/9899/\"", "rc": 1, "failed": true, "invocation": {"module_name": "synchronize", "module_args": {"dest": "/tmp/9899/", "src": "roles/", "rsync_opts": "--chmod='u=rwX,g=rX'"}}, "msg": "rsync: Invalid argument passed to --chmod (u=rwX g=rX)\nrsync error: syntax or usage error (code 1) at main.c(1554) [client=3.1.1]\n"}
</pre>

The issue being the permissions portion is being split on the comma and rejoined incorrectly. Removing the quotes around the permissions resolves the issue (`--chmod u=rwX,g=rX`), so there is still apparently incorrect in the parsing of these options.

**Comment by abadger** [comment]
Took a look a this and the module is actually documented to take a list:

```
- rsync_opts
        Specify additional rsync options by passing in an array.
        [Default: None]
```

Using a list for the rsync_opts value makes this work with and without quotes.  (The example shows a string,  I'll fix that momentarily)

Because rsync_opts is a list, attempting to make it handle quotes would be a backwards incompatible change to all modules whose parameters are a list and may exclude some use cases (where a list parameter to a module would actually need to contain quotes).  So I'll make that change to the documentation mentioned above and people should just use the list format to get this to work correctly.

**Comment by abadger** [comment]
Documentation fixed here: https://github.com/ansible/ansible-modules-core/commit/e9139ba4fa66f4f632ca9b200f3abf277a0e6ca4

So this should all be fixed in devel (which will be 2.0) now!  Thanks for reporting this problem.

# Closing This Ticket

Hi!

We believe recent commits (likely detailed above) should resolve this question or problem for you.

This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
